libvast/src/format/syslog.cpp

```diff
@@ -12,9 +12,85 @@
  ******************************************************************************/
 
 #include "vast/format/syslog.hpp"
+#include "vast/error.hpp"
 
-namespace vast::format::mrt {
+namespace vast {
+namespace format {
+namespace syslog {
 
-// TODO: implement functions here that do not belong into the header file. 
+syslog_parser::syslog_parser() {
+  // Syslog header type.
+  auto header_fields = std::vector<record_field>{
+    {"facility", count_type{}},
+    {"severity", count_type{}},
+    {"time", time_type{}.attributes({{"timestamp"}})},
+    {"hostname", string_type{}},
+    {"app_name", string_type{}},
+    {"process_id", string_type{}},
+    {"message_id", string_type{}},
+    {"version", count_type{}},
+    
+  };
+  header_type = record_type{std::move(header_fields)}.name("syslog.header");
+  
+  // Syslog structured_data type.
+  // There can be multiple key value pairs for one structured data element
+  // Is it possible to store a key-value-pair in one record OR is it better to define for the separated record fields?
+  auto structured_data_fields = std::vector<record_field>{
+    {"sd_id", string_type{}},
+    {"sd_elements", string_type{}},
+  }; 
+  structured_data_type = record_type{std::move(structured_data_fields)}.name(
+    "syslog.strucured_data");
 
-} // namespace vast::format::mrt
+  // Syslog message content type.
+  auto message_content_fields = std::vector<record_field>{
+    {"message", string_type{}}
+  };
+  message_type = record_type{std::move(message_content_fields)}.name(
+    "syslog.message");
+}
+
+caf::error reader::schema(vast::schema x) {
+  auto xs = {
+    &parser_.header_type,
+    &parser_.structured_data_type,
+    &parser_.message_type,
+  };
+  return replace_if_congruent(xs, x);
+}
+
+vast::schema reader::schema() const {
+  vast::schema sch;
+  sch.add(parser_.header_type);
+  sch.add(parser_.structured_data_type);
+  sch.add(parser_.message_type);
+  return sch;
+}
+
+const char* reader::name() const {
+  return "syslog-reader";
+}
+
+
+// Here I am unsure if I have the correct idea, I looked into pcap, zeek and mrt to try to understand the reader logic
+caf::error reader::read_impl(size_t max_events, size_t max_slice_size,
+                             consumer& f) {
+  // Sanity checks.
+  VAST_ASSERT(max_events > 0);
+  VAST_ASSERT(max_slice_size > 0);
+  // EOF check.
+  if (lines_->done())
+    return make_error(ec::end_of_input, "input exhausted");
+  
+  for (size_t produced = 0; produced < max_events; ++produced) {
+    if (builder_->rows() == max_slice_size)
```

Comment:
Before this, you need to run the parser and create table slices for the current input line.